ddtrace/_events.py

```diff
@@ -0,0 +1,112 @@
+from typing import Callable
+from typing import Mapping
+from typing import Optional
+from typing import Type
+from typing import Union
+
+import attr
+
+from ddtrace import Span
+from ddtrace import _hooks
+from ddtrace import config
+from ddtrace.internal import compat
+
+
+_HOOKS = _hooks.Hooks()
+
+
+@attr.s(frozen=True)
+class IntegrationEvent(object):
+    """
+    An IntegrationEvent is emitted by an integration (e.g. the flask framework integration)
+    and is linked to a span.
+    """
+    span = attr.ib(type=Span)
+    integration = attr.ib(type=str)
+
+    def emit(self):
+        """Alias for emitting this event."""
+        emit(self)
+
+    @classmethod
+    def register(cls, func=None, integration=None):
+        """Alias for registering a listener for this event type."""
+        return register(cls, func=func, integration=integration)
+
+    @classmethod
+    def deregister(cls, func, integration=None):
+        """Alias for deregistering a listener for this event type."""
+        deregister(cls, func, integration=integration)
+
+    if config._raise:
+        @span.validator  # type: ignore
+        def check_span(self, attribute, value):
+            assert isinstance(value, Span)
+
+        @integration.validator  # type: ignore
+        def check_integration(self, attribute, value):
+            assert value in config._config
+
+
+@attr.s(frozen=True)
+class WebRequest(IntegrationEvent):
+    """
+    The WebRequest event is emitted by web framework integrations before the WebResponse event.
+    """
+    method = attr.ib(type=str)
+    url = attr.ib(type=str)
+    headers = attr.ib(type=Mapping[str, str], factory=dict)
```

Comment:
I'm going to be picky here, but the type is wrong AFAIK.
You can have multiple values for a header, so `[str, str]` is a limited approach.
I'd rather be `Mapping[str, Iterable[str]]` or something like that.

Reply:
💯 agreed! I will apply the changes.